src/probability_theory/stopping.lean

```diff
@@ -0,0 +1,294 @@
+end
+
+lemma is_stopping_time_const {f : filtration ι m} (i : ι) :
+  is_stopping_time f (λ x, i) :=
+λ j, by simp
+
+end preorder
+
+namespace is_stopping_time
+
+lemma max [linear_order ι] {f : filtration ι m} {τ π : α → ι}
+  (hτ : is_stopping_time f τ) (hπ : is_stopping_time f π) :
+  is_stopping_time f (λ x, max (τ x) (π x)) :=
+begin
+  intro i,
+  simp_rw [max_le_iff, set.set_of_and],
+  exact @measurable_set.inter _ (f i) _ _ (hτ i) (hπ i),
+end
+
+lemma min [linear_order ι] {f : filtration ι m} {τ π : α → ι}
+  (hτ : is_stopping_time f τ) (hπ : is_stopping_time f π) :
+  is_stopping_time f (λ x, min (τ x) (π x)) :=
+begin
+  intro i,
+  simp_rw [min_le_iff, set.set_of_or],
+  exact @measurable_set.union _ (f i) _ _ (hτ i) (hπ i),
+end
+
+lemma add_const
+  [add_group ι] [preorder ι] [covariant_class ι ι (function.swap (+)) (≤)]
+  [covariant_class ι ι (+) (≤)]
+  {f : filtration ι m} {τ : α → ι} (hτ : is_stopping_time f τ) {i : ι} (hi : 0 ≤ i) :
+  is_stopping_time f (λ x, τ x + i) :=
+begin
+  intro j,
+  simp_rw [← le_sub_iff_add_le],
+  exact f.mono (sub_le_self j hi) _ (hτ (j - i)),
+end
+
+variables [preorder ι] {f : filtration ι m}
+
+/-- The associated σ-algebra with a stopping time. -/
+def measurable_space
+  {τ : α → ι} (hτ : is_stopping_time f τ) : measurable_space α :=
+{ measurable_set' := λ s, ∀ i : ι, measurable_set[f i] (s ∩ {x | τ x ≤ i}),
+    measurable_set_empty :=
+      λ i, (set.empty_inter {x | τ x ≤ i}).symm ▸ @measurable_set.empty _ (f i),
+    measurable_set_compl := λ s hs i,
+      begin
+        rw (_ : sᶜ ∩ {x | τ x ≤ i} = (sᶜ ∪ {x | τ x ≤ i}ᶜ) ∩ {x | τ x ≤ i}),
+        { refine @measurable_set.inter _ (f i) _ _ _ _,
+          { rw ← set.compl_inter,
+            exact @measurable_set.compl _ _ (f i) (hs i) },
+          { exact hτ i} },
+        { rw set.union_inter_distrib_right,
+          simp only [set.compl_inter_self, set.union_empty] }
+      end,
+    measurable_set_Union := λ s hs i,
+      begin
+        rw forall_swap at hs,
+        rw set.Union_inter,
+        exact @measurable_set.Union _ _ (f i) _ _ (hs i),
+      end }
+
+lemma measurable_set {τ : α → ι} (hτ : is_stopping_time f τ) (s : set α) :
+  measurable_set[hτ.measurable_space] s ↔
+  ∀ i : ι, measurable_set[f i] (s ∩ {x | τ x ≤ i}) :=
+iff.rfl
+
+lemma measurable_space_mono
+  {τ π : α → ι} (hτ : is_stopping_time f τ) (hπ : is_stopping_time f π) (hle : τ ≤ π) :
+  hτ.measurable_space ≤ hπ.measurable_space :=
+begin
+  intros s hs i,
+  rw (_ : s ∩ {x | π x ≤ i} = s ∩ {x | τ x ≤ i} ∩ {x | π x ≤ i}),
+  { exact @measurable_set.inter _ (f i) _ _ (hs i) (hπ i) },
+  { ext,
+    simp only [set.mem_inter_eq, iff_self_and, and.congr_left_iff, set.mem_set_of_eq],
+    intros hle' _,
+    exact le_trans (hle _) hle' },
+end
+
+lemma measurable_space_le [encodable ι]
+  {τ : α → ι} (hτ : is_stopping_time f τ) :
+  hτ.measurable_space ≤ m :=
+begin
+  intros s hs,
+  change ∀ i, measurable_set[f i] (s ∩ {x | τ x ≤ i}) at hs,
```

Comment:
Is that `change` the same as this?
```suggestion
  rw is_stopping_time.measurable_set at hs,
```
Also: the lemma named `measurable_set` should be `protected`

Reply:
`rw` doesn't work since we don't have syntactical equality here as we have  `measurable_set'` instead of `measurable_set`